src/auth/passkey-config.ts

```diff
@@ -0,0 +1,188 @@
+/*!
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as validator from '../utils/validator';
+import { AuthClientErrorCode, FirebaseAuthError } from '../utils/error';
+import { deepCopy } from '../utils/deep-copy';
+
+/**
+ * Interface representing the properties to update in a passkey config.
+ */
+export interface PasskeyConfigRequest {
+    /**
+     * An array of website or app origins. Only challenges signed
+     * from these origins will be allowed for signing in with passkeys.
+     */
+    expectedOrigins?: string[];
+}
+
+/**
+ * Response received from the server when retrieving, creating, or updating the passkey config.
+ */
+export interface PasskeyConfigServerResponse {
+    name?: string;
+    rpId?: string;
+    expectedOrigins?: string[];
+}
+
+/**
+ * Request for creating or updating the passkey config on the server.
+ */
+export interface PasskeyConfigClientRequest {
+    rpId?: string;
+    expectedOrigins?: string[];
+}
+
+/**
+ * Configuration for signing in users using passkeys.
+ */
+export class PasskeyConfig {
+  /**
+   * The name of the PasskeyConfig resource.
+   */
+  public readonly name?: string;
+  /**
+   * The relying party ID for passkey verifications.
+   * This cannot be changed once created.
+   */
+  public readonly rpId?: string;
+  /**
+   * The allowed website or app origins.
+   * Only challenges signed from these origins will be allowed for signing in with passkeys.
+   */
+  public readonly expectedOrigins?: string[];
+
+  /**
+   * Validates a passkey config request object and throws an error on failure.
+   * @param isCreateRequest - A boolean indicating if it's a create request or not.
+   * @param passkeyConfigRequest - Passkey config to be set.
+   * @param rpId - (optional) Relying party ID if it's a create request.
+   * @throws FirebaseAuthError - If validation fails.
+   * 
+   * @internal
+   */
+  private static validate(isCreateRequest: boolean, passkeyConfigRequest?: PasskeyConfigRequest, rpId?: string): void {
+    // Validation for creating a new PasskeyConfig.
+    if (isCreateRequest && !validator.isNonEmptyString(rpId)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        "'rpId' must be a non-empty string.",
+      );
+    }
+    // Validation for updating an existing PasskeyConfig.
+    if (!isCreateRequest && typeof rpId !== 'undefined') {
```

Comment:
do you allow customer to pass in the existing value rpId for updates?

Reply:
Doesn't that seem a little redundant? Since we can't change rpId once created?